Connections

ThatOneSix opened the pull request:
### Change Scope

* This change adds a new container, 'sockets', under /network-instance/. It is loosely based on the `netstat` command. There is room to expand the container at a later date to include other sub-containers as needed.
* This change is backward compatible.

### Platform Implementations

 * [Ubuntu `netstat` manual](https://manpages.ubuntu.com/manpages/jammy/en/man8/netstat.8.html) 
 * [Aruba `show datapath session table`](https://arubanetworking.hpe.com/techdocs/CLI-Bank/Content/sd-branch/sh-dp-sestbl.htm)
 
```
augment /oc-ni:network-instances/oc-ni:network-instance:
    +--rw sockets
       +--ro socket* [local-address local-port remote-address remote-port]
          +--ro local-address     -> ../state/local-address
          +--ro local-port        -> ../state/local-port
          +--ro remote-address    -> ../state/remote-address
          +--ro remote-port       -> ../state/remote-port
          +--ro state
             +--ro socket-type?      union
             +--ro local-address?    oc-inet:host
             +--ro local-port?       oc-inet:port-number
             +--ro remote-address?   oc-inet:host
             +--ro remote-port?      oc-inet:port-number
             +--ro tcp-state?        enumeration
             +--ro process-name?     string
             +--ro pid?              uint64
```

dplore commented:
/gcbrun

dplore commented:
/gcbrun

ThatOneSix commented:
Changed to use `augment` in the sub-module and removed all changes to openconfig-system.yang.

dplore commented:
/gcbrun

jsterne commented:
Don't we likely need a network-instance dimension to this?

ThatOneSix commented:
To clarify, do you mean adding something like...

```
import openconfig-network-instance { prefix oc-ni; }

leaf network-instance {
      type oc-ni:network-instance-ref;
      description
        "Lorem ipsum";
```

So that, for example, L3 devices can report which VRF the connection is on?

ThatOneSix commented:
Oh, I didn't consider that. Good point. So something like...

+--rw system
|  +--rw connections
|  |  +--rw network-instance [network-instance]
|  |  |  +--rw network-sockets
|  |  |      +--rw network-socket [socket]

Where socket is source IP + port.

Or

+--rw system
|  +--rw connections
|  |  +--rw internet-sockets
|  |     +--rw internet-socket [socket]

Where socket is IP + port + network instance.

Please correct me if I've misunderstood.

Looking at the two, I think the first makes more sense.

jsterne commented:
I somewhat prefer the first one as well.

Another question: aren't sockets sometimes identified by a 4-part tuple (local ip, local port, remote ip, remote port)?  [listening ports are maybe just local ip + local port]

ThatOneSix commented:
I've reworked the file to the following:

module: openconfig-connections

  augment /oc-sys:system:
    +--rw connections
       +--rw network-instances
          +--ro network-instance* [instance]
             +--ro instance           -> ../state/instance
             +--ro state
             |  +--ro instance?   oc-ni:network-instance-ref
             +--ro network-sockets
                +--ro network-socket* [socket]
                   +--ro socket    -> ../state/socket
                   +--ro state
                      +--ro socket?           string
                      +--ro protocol?         union
                      +--ro local-address?    oc-inet:host
                      +--ro local-port?       oc-inet:port-number
                      +--ro remote-address?   oc-inet:host
                      +--ro remote-port?      oc-inet:port-number
                      +--ro tcp-state?        enumeration
                      +--ro process-name?     string
                      +--ro pid?              uint64

Ran through pyang lint successfully.

As for the socket, I was mostly using local address + local port to make it more readable. I am open to changing the key for the socket list to something else.

earies commented:
RE: `netstat` - note netstat is obsolete, suggest referring to iproute2 `ss` and `ip` utils rather as an alternative

Structural comment based off prior comments....

If we are going to enumerate network-instances, I'd propose we not create yet another list but rather this could be anchored under `/network-instances`

But when it comes to `/system` we often have the design decision to choose to treat the system as a single entity which then often means "active controller-card" only or consider the possibility of a distributed system where connections, pids, storage, etc... could all be a subset of a component thus possibly warranting this to be a subset of `/components` (We already see this w/ prior published system related items)

Another quick comment as to the naming of the module and top-level node - imo "connections" is too vague here so suggest narrowing this in rather to the categories that correspond to the above utilities (if we lean towards such which I think is mostly ok for this purpose)

earies commented on release/models/system/openconfig-connections.yang:
```diff
@@ -0,0 +1,291 @@
+    }
+
+    leaf pid {
+      type uint64;
```
While `openconfig-procmon` defines a pid as `uint64`, I think this is too generous.  Linux/BSD as well as VxWorks based systems use `uint32` for PIDs/tasks often w/ a much lower max/default

```
$ cat /proc/sys/kernel/pid_max 
32768
```

dplore commented:
I suspect to resolve the issue about network-instances, adding it as part of the "connections" list key simplifies the model and matches the linux implementation well.  


```
module: openconfig-connections

  augment /oc-sys:system:
    +--rw connections
       +--rw connection* [network-instance,socket]
             +--ro network-instance -> ../state/network-instance
             +--ro socket           -> ../state/socker
             +--ro state
             |  +--ro network-instance?   oc-ni:network-instance-ref
             |  +--ro socket?           string
             |  +--ro protocol?         union
                +--ro local-address?    oc-inet:host
                +--ro local-port?       oc-inet:port-number
                +--ro remote-address?   oc-inet:host
                +--ro remote-port?      oc-inet:port-number
                +--ro tcp-state?        enumeration
                +--ro process-name?     string
                +--ro pid?              uint64
```


Regarding multiple CONTROLLER_CARDs, we can choose (in a future PR) to replicate this as a subtree underneath /components/component when type = CONTROLLER_CARD.  ie:
`/components/component/controller-card/connections/connection`

This follows the style that /system is a copy of the current state of the "whole device" and /components provides a "drill down" view into individual component state.

ThatOneSix commented:
Modified the openconfig-connections.yang file to incorporate suggestions about simplifying the list. Ran through pyang lint and received no errors. The tree it output is:

```
module: openconfig-system
  +--rw system
     +--rw oc-conn:connections
        +--ro oc-conn:connection* [network-instance socket]
           +--ro oc-conn:network-instance    -> ../state/network-instance
           +--ro oc-conn:socket              -> ../state/socket
           +--ro oc-conn:state
              +--ro oc-conn:socket?             string
              +--ro oc-conn:network-instance?   oc-ni:network-instance-ref
              +--ro oc-conn:protocol?           union
              +--ro oc-conn:local-address?      oc-inet:host
              +--ro oc-conn:local-port?         oc-inet:port-number
              +--ro oc-conn:remote-address?     oc-inet:host
              +--ro oc-conn:remote-port?        oc-inet:port-number
              +--ro oc-conn:tcp-state?          enumeration
              +--ro oc-conn:process-name?       string
              +--ro oc-conn:pid?                uint64
```

dplore commented:
/gcbrun

dplore commented:
/gcbrun

LimeHat reviewed (changes requested):
1. A major issue with the proposal is the definition of what we are trying to expose.
If we are talking about _connections_, then this applies only to TCP. Most of the other standard protocols (IP, UDP, ICMP) are connectionless by definition.
I'm not sure what is the intention: to list the sockets, or to list the connections, but the current combination of the proposed models (that lists many protocols) and the terminology/description (connections, active connections) doesn't make much sense to me.
2. I would question the placement of this in /system/ tree. IMO a better home for this is /network-instances/ tree (where each NI can expose its own state) -> I don't know if this was discussed in earlier calls, but to me this would be a much cleaner model.

LimeHat commented on release/models/system/openconfig-connections.yang:
```diff
@@ -0,0 +1,247 @@
+            description
+              "Index value indicating TCP";
+          }
+          enum TCP6 {
```
Does this mean that the other TCP (above) is TCP over IPv4? (the description doesn't mention that)
What's the motivation for this split?

ThatOneSix replied:
As mentioned in the description, this is based on the `netstat` command, which separates TCPv4 and TCPv6 traffic into `TCP` and `TCP6`. Depending on the vendor implementation, they could choose to use TCP for both IPv4 and IPv6 traffic, or split them into the separate enums. I also included `type string;` at the end of the enums so that vendors could use additional/different descriptions as they prefer.

ghananigans replied:
+1 to what @LimeHat said.

What problem are we trying to solve with this model change? 

I don't think there is an issue using existing solutions like `netstat`/`/proc/net`/`sock_diag` for inspiration when solving our problems, but we should also confirm how our problems align with those of existing solutions before doing so.

ghananigans commented on release/models/system/openconfig-connections.yang:
```diff
@@ -0,0 +1,247 @@
+        instance is specified, DEFAULT_INSTANCE is used.";
+    }
+
+    leaf protocol {
```
This seems to read like less like a protocol and more like a socket _type_.

ghananigans replied:
"socket-type" sgtm!

ghananigans commented on release/models/system/openconfig-connections.yang:
```diff
@@ -0,0 +1,247 @@
+        description
+          "Parameters related to individual active network connections";
+
+        leaf network-instance {
```
What does this leaf point to? In the context of Ubuntu/Linux (since that is 1 of 2 implementation listed in the PR description) does it point to the network interface that a socket is _explicitly_ bound to, or the interface a route for the socket is cached to egress through? 

What if the socket is bound to a specific local address but no remote, what should show up here? Similarly, what if the socket is bound all addresses?

ghananigans commented on release/models/system/openconfig-connections.yang:
```diff
@@ -0,0 +1,247 @@
+
+  // grouping statements
+
+  grouping connections-top {
```
As @LimeHat suggested, I don't think "connections" is an appropriate name for a container that includes connection-_less_ protocols/socket.

ThatOneSix replied:
For what it's worth, even the `ss` manual describes connectionless protocols as having established a connection.

"When no option is used ss displays a list of open non-listening sockets (e.g. TCP/UNIX/UDP) that have established connection."

https://man7.org/linux/man-pages/man8/ss.8.html

LimeHat commented:
> 1. A major issue with the proposal is the definition of what we are trying to expose.
>    If we are talking about _connections_, then this applies only to TCP. Most of the other standard protocols (IP, UDP, ICMP) are connectionless by definition.
>    I'm not sure what is the intention: to list the sockets, or to list the connections, but the current combination of the proposed models (that lists many protocols) and the terminology/description (connections, active connections) doesn't make much sense to me.
> 2. I would question the placement of this in /system/ tree. IMO a better home for this is /network-instances/ tree (where each NI can expose its own state) -> I don't know if this was discussed in earlier calls, but to me this would be a much cleaner model.

I would like to get my comments addressed as well.

dplore commented:
> > 1. A major issue with the proposal is the definition of what we are trying to expose.
> >    If we are talking about _connections_, then this applies only to TCP. Most of the other standard protocols (IP, UDP, ICMP) are connectionless by definition.
> >    I'm not sure what is the intention: to list the sockets, or to list the connections, but the current combination of the proposed models (that lists many protocols) and the terminology/description (connections, active connections) doesn't make much sense to me.
> > 2. I would question the placement of this in /system/ tree. IMO a better home for this is /network-instances/ tree (where each NI can expose its own state) -> I don't know if this was discussed in earlier calls, but to me this would be a much cleaner model.
> 
> I would like to get my comments addressed as well.

`/system` is a good location for this.  `/system` is intended to represent the state of the device's currently active running operating system and is mostly independent of all the "networking" configuration.  

Because the intent of `/system` is to model state of the NOS and Linux is the dominant kernel used and is open, it's precedent is strong.   While the UDP protocol is connectionless and does not track a connection state, the "system" does. This PR models system state, not the packets on the wire.

Regarding `connections`, your comments are very insightful and appreciated.  Maybe this clarification of modeling the "system" helps explain why `connections` is defensible.  If you still feel otherwise, then in order to affect this PR please provide some alternative which can be reviewed.  The suggestion will need stronger and more relevant precedent than what is already provided.  The final decision on the naming is up to the rough consensus of the OC Operators. 

Expanding the scope to include listening ports is at the discretion of PR author.  It sounds like a good idea.  But if there isn't an operational need, modeling listening sockets doesn't have to be included in this PR at this time.

LimeHat commented:
> This PR models system state, not the packets on the wire.

I totally agree with that, but that doesn't mean it belongs to the `/system/` :-)

> /system is intended to represent the state of the device's currently active running operating system and is mostly independent of all the "networking" configuration.

Great, then this is exactly why the connections should be moved to `/network-instance`.

Connections and sockets are not independent from the networking configuration. Network instances are (typically) mapped to [network namespaces](https://man7.org/linux/man-pages/man7/network_namespaces.7.html) in Linux! Each network namespace operates its own set of interfaces and sockets. They are totally configuration dependent. If i create network-instances `mgmt1` and `mgmt2`, I will create two new network namespaces in linux. Each namespace will have its own socket table (`/proc/net/udp`, `/proc/net/tcp` that are consumed by netstat), etc. There's nothing global about this :-) And sockets and connections will be opened according to configuration of apps that attached to these network instances.

ghananigans commented:
> > This PR models system state, not the packets on the wire.
> 
> I totally agree with that, but that doesn't mean it belongs to the `/system/` :-)
> 
> > /system is intended to represent the state of the device's currently active running operating system and is mostly independent of all the "networking" configuration.
> 
> Great, then this is exactly why the connections should be moved to `/network-instance`.
> 
> Connections and sockets are not independent from the networking configuration. Network instances are (typically) mapped to [network namespaces](https://man7.org/linux/man-pages/man7/network_namespaces.7.html) in Linux! Each network namespace operates its own set of interfaces and sockets. They are totally configuration dependent. If i create network-instances `mgmt1` and `mgmt2`, I will create two new network namespaces in linux. Each namespace will have its own socket table (`/proc/net/udp`, `/proc/net/tcp` that are consumed by netstat), etc. There's nothing global about this :-) And sockets and connections will be opened according to the configuration of apps that attached to these network instances.

Thats a good point, @LimeHat. Thank you for pointing that out. I hadn't considered the fact that, at least on Linux, each network namespace maintains its own distinct set of socket tables and to support a full query of sockets on a system, we would have to check each namespace. 

It's unclear to me if _all_ vendors implement sockets & network-instances this way, but I think expressing this with the model is fine, so long as a socket cannot easily/non-disruptively move across network-instances throughout their lifetime (I believe this to be true on most platforms, but just explicitly pointing it out as an assumption I have).

```
# Terminal 1
$ sudo ip netns exec testing nc -u -l :: 12346 

# Terminal 2
$ nc -u -l :: 12345

# Terminal 3
$ netstat -tunlp | grep "12345\|12346"
udp6       0      0 :::12345                :::*                                796206/nc           
$ sudo ip netns exec testing netstat -tunlp | grep "12345\|12346"
udp6       0      0 :::12346                :::*                                796592/nc       
```
Given the above, I now _do_ believe that sockets may be better placed within `network-instance` which itself may be part a wider platform/system alongside other `network-instances`.

dplore commented:
@ThatOneSix I'd like to see a response regarding modeling socket connections and namespaces.  Are sockets in linux contained within namespaces?  How do the vendor implemenations you want to support this compare?  Do they support vrfs/network-instances?

ThatOneSix commented:
@dplore Apologies for the delay; I was out of the country. I will review and respond ASAP.

ThatOneSix commented:
@dplore @LimeHat @ghananigans

I have updated the sockets config to now be in network-instances. No issues when run through pyang lint. This is the tree:

```
+--rw oc-sockets:sockets
           +--ro oc-sockets:socket* [socket]
              +--ro oc-sockets:socket    -> ../state/socket
              +--ro oc-sockets:state
                 +--ro oc-sockets:socket?           string
                 +--ro oc-sockets:socket-type?      union
                 +--ro oc-sockets:local-address?    oc-inet:host
                 +--ro oc-sockets:local-port?       oc-inet:port-number
                 +--ro oc-sockets:remote-address?   oc-inet:host
                 +--ro oc-sockets:remote-port?      oc-inet:port-number
                 +--ro oc-sockets:tcp-state?        enumeration
                 +--ro oc-sockets:process-name?     string
                 +--ro oc-sockets:pid?              uint64
```

I have modified it to fit other suggestions as well, such as removing TCP6 and UDP6 and making the language focus on sockets instead of connections (other than TCP state info).

ThatOneSix commented:
Added the new line characters at the ends of the YAML files to fix the lint issue.

dplore commented:
/gcbrun
